### PR TITLE
Copter/Global: Console output can be disabled

### PR DESCRIPTION
**Pull request by muramura**
I have a USB connection between FC and GCS.
I debugged on the console when I was NUTTX.
I have not used the console since CHIBIOS.
I use HAL.UART for software debugging.
My drone does not use the messages output to the console.
I do not know how to use this message and consider it useless.
I can add a preprocessor for HAL_CONSOLE_DISABLED and choose not to output to the console.

If you do not use this message, you can free up about 4 Kbytes.

### CODE REVIEW

File: libraries/AP_SBusOut/AP_SBusOut.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ void AP_SBusOut::init() {
     uint16_t rate = sbus_rate.get();
 
 #if SBUS_DEBUG
-    hal.console->printf("AP_SBusOut: init %d Hz\n", rate);
+    ("AP_SBusOut: init %d Hz\n", rate);
```
Comment (hendjoshsr71) [inline libraries/AP_SBusOut/AP_SBusOut.cpp:174]:
```suggestion
    PRINTF("AP_SBusOut: init %d Hz\n", rate);
```

File: libraries/AP_SBusOut/AP_SBusOut.cpp [hunk 2]
```diff
@@ -99,7 +99,7 @@ void AP_SBusOut::sbus_format_frame(uint16_t *channels, uint8_t num_channels, uin
         static uint16_t lastch9 = 0;
         if ((i==8) && (pwmval != lastch9)) {
             lastch9 = pwmval;
-            hal.console->printf("channel 9 pwm: %04d\n", pwmval);
+            ("channel 9 pwm: %04d\n", pwmval);
```
[the remark this answers is not in the document]
Comment (muramura) [reply libraries/AP_SBusOut/AP_SBusOut.cpp]: I made a mistake.
I'll get back to it.
This is intended to output when SBUS_DEBUG is enabled.

File: libraries/AP_SBusOut/AP_SBusOut.cpp (hunk 1 above)
Comment (muramura) [reply libraries/AP_SBusOut/AP_SBusOut.cpp:174]: I made a mistake.
I'll get back to it.
This is intended to output when SBUS_DEBUG is enabled.

**Comment by HefnySco** [comment]
When I define 
#define HAL_CONSOLE_DISABLED
it gives me this:
 
image

**Comment by HefnySco** [comment]
Also what is the purpose of do..while statement.
It works fine without it.

image

Thanks

**Comment by khancyr** [comment]
@HefnySco the do while allow to enclose the macro statement. It is better to have it to prevent error in futur

**Comment by tridge** [comment]
I think we need the name to be more distinctive, to prevent conflicts with include files. Maybe DEV_PRINTF() ?

**Review by rmackay9 — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 3]
```diff
@@ -1156,7 +1156,7 @@ void UARTDriver::write_pending_bytes(void)
         }
         if (AP_HAL::micros() - _first_write_started_us > 500*1000UL) {
             // it doesn't look like hw flow control is working
-            hal.console->printf("disabling flow control on serial %u\n", sdef.get_index());
+            DEV_PRINTF("disabling flow control on serial %u\n", sdef.get_index());
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:1159]: Delete

```suggestion
```

File: libraries/AP_Compass/AP_Compass_LSM9DS1.cpp [hunk 4]
```diff
@@ -84,23 +84,23 @@ bool AP_Compass_LSM9DS1::init()
...
     if (!bus_sem) {
-        hal.console->printf("LSM9DS1: Unable to get bus semaphore\n");
+        DEV_PRINTF("LSM9DS1: Unable to get bus semaphore\n");
         return false;
     }
     bus_sem->take_blocking();
 
     if (!_check_id()) {
-        hal.console->printf("LSM9DS1: Could not check id\n");
+        DEV_PRINTF("LSM9DS1: Could not check id\n");
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_LSM9DS1.cpp:93]: example to remove and make a comment instead
```suggestion
```

File: libraries/AP_Compass/AP_Compass_LSM9DS1.cpp [hunk 5]
```diff
@@ -84,23 +84,23 @@ bool AP_Compass_LSM9DS1::init()
...
         goto errout;
     }
 
     if (!_configure()) {
-        hal.console->printf("LSM9DS1: Could not check id\n");
+        DEV_PRINTF("LSM9DS1: Could not check id\n");
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_LSM9DS1.cpp:98]: example to remove

```suggestion
```

File: libraries/AP_Compass/AP_Compass_RM3100.cpp [hunk 6]
```diff
@@ -149,7 +149,7 @@ bool AP_Compass_RM3100::init()
     }
     set_dev_id(compass_instance, dev->get_bus_id());
 
-    hal.console->printf("RM3100: Found at address 0x%x as compass %u\n", dev->get_bus_address(), compass_instance);
+    DEV_PRINTF("RM3100: Found at address 0x%x as compass %u\n", dev->get_bus_address(), compass_instance);
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_RM3100.cpp:152]: Example to remove of normal status text that isn't very useful

```suggestion

```

**Review by hendjoshsr71 — COMMENTED** [review]
DevCall Discussion for follow ups:

Consider going through a couple libraries: Capture around a 100 of these and see about removing them in a different PR and any associated code.

Consider switching them to a per library basis for ones we might want to keep. 

Not a good idea to globally define them all out even for Periph. Also not a good idea to expand these to GCS_SEND_TEXT either since then users may be get worried by the extra messaging. 

Don't have to switch all of them over that weren't captured here.

**Comment by hendjoshsr71** [comment]
No size difference at all on any board. Though interestingly the hash check says they are different?
image

marking MergeOnCI pass since if they were different the size should have changed?

**Comment by HefnySco** [comment]
> No size difference at all on any board. Though interestingly the hash check says they are different? image
> 
> marking MergeOnCI pass since if they were different the size should have changed?

I believe this is because it is never removed,  it is replaced by the [do...while](https://github.com/ArduPilot/ardupilot/pull/20340#issuecomment-1075800962) which I believe should be removed in the optimization stage by the compiler.

**Comment by khancyr** [comment]
elf_diff show no difference, so something fishy appear with checksum, maybe a path things in debug variable.

**Pull request merged**